### PR TITLE
Add SCryptPasswordEncoder

### PR DESCRIPTION
**Pull request by shazin**
I have signed and agree to the terms of the SpringSource Individual Contributor License Agreement.

### CODE REVIEW

**Comment by rwinch** [comment]
Thanks for the PR!

**Comment by rwinch** [comment]
@shazin Looking at the referenced library with a bit more scrutiny, I'm a bit hesitant to use it. Among other things, the library has not had a commit in 3 years which is a bit scary for me.

It might be nice to explore alternatives. For example, [Bouncy Castle's SCrypt](http://bouncy-castle.1462172.n4.nabble.com/Java-Bouncy-Castle-scrypt-implementation-td4656832.html)

**Comment by rwinch** [comment]
This resolves #3702

**Comment by shazin** [comment]
Thanks for the review @rwinch. Will make the changes and push.

**Comment by rwinch** [comment]
@shazin Thanks for the fast response. Another thing to point out is that the dependencies should go into the crypto.gradle (Maven poms are generated) and should be optional.

**Comment by shazin** [comment]
Changes with Bouncy Castle Implementation of SCrypt pushed.

**Comment by rwinch** [comment]
@shazin Thanks for the updates!

File: crypto/crypto.gradle [hunk 1]
```diff
@@ -10,3 +10,7 @@ configure(project.tasks.withType(Test)) {
 		exclude '**/EncryptorsTests.class'
 	}
 }
+
+dependencies {
+    compile 'org.bouncycastle:bcprov-jdk15on:1.54'
```
Comment (rwinch) [inline crypto/crypto.gradle]: This should be optional. Also don't forget to regenerate the pom.xml using the following from this directory

../gradlew generatePom

**Comment by rwinch** [comment]
Please make sure you use tabs not spaces (there are some spaces in the files)

File: crypto/src/test/java/org/springframework/security/crypto/scrypt/SCryptPasswordEncoderTests.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.crypto.scrypt;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
```
Comment (rwinch) [inline crypto/src/test/java/org/springframework/security/crypto/scrypt/SCryptPasswordEncoderTests.java]: Please use assertj for your assertions. 

See http://joel-costigliola.github.io/assertj/

File: crypto/src/main/java/org/springframework/security/crypto/scrypt/SCryptPasswordEncoder.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    private final int cpuCost;
+    
+    private final int memoryCost;
+    
+    private final int parallelization;
+    
+    private final byte[] saltBytes;
+    
+    public SCryptPasswordEncoder() {
+        this(new String(KeyGenerators.secureRandom().generateKey()), 16384, 8, 1);
```
Comment (rwinch) [inline crypto/src/main/java/org/springframework/security/crypto/scrypt/SCryptPasswordEncoder.java]: Where did you get your default values? What do they imply (i.e. how much memory is used, how much time does it take on your CPU, etc)?

File: crypto/src/main/java/org/springframework/security/crypto/scrypt/SCryptPasswordEncoder.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    }
+    
+    /**
+     * @param salt value for the algorithm
+     * @param cpu cost of the algorithm. must be power of 2 greater than 1
+     * @param memory cost of the algorithm
+     * @param parallelization of the algorithm
+     */
+    public SCryptPasswordEncoder(CharSequence salt, int cpuCost, int memoryCost, int parallelization) {
```
Comment (rwinch) [inline crypto/src/main/java/org/springframework/security/crypto/scrypt/SCryptPasswordEncoder.java]: Salt should be secure random generated for each individual password. That means we should not pass it into the constructor. Instead we should allow for a salt length to be injected and the salt should be added as a prefix of the password. Something like:

<salt><password>

we can then verify the password by extracting the salt using the number of bytes to extract out the salt that was generated.

File: crypto/src/main/java/org/springframework/security/crypto/scrypt/SCryptPasswordEncoder.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        
+        this.saltBytes = Utf8.encode(salt);
+        this.cpuCost = cpuCost;
+        this.memoryCost = memoryCost;
+        this.parallelization = parallelization;
+    }
+
+	@Override
+	public String encode(CharSequence rawPassword) {
+        return new String(Hex.encode(SCrypt.generate(Utf8.encode(rawPassword), saltBytes, cpuCost, memoryCost, parallelization, saltBytes.length)));
```
Comment (rwinch) [inline crypto/src/main/java/org/springframework/security/crypto/scrypt/SCryptPasswordEncoder.java]: Ideally we should be encoding with the parameters in the encoded value (along with the salt as mentioned above). This allows for increasing the strength as computing resources get better.

File: crypto/src/test/java/org/springframework/security/crypto/scrypt/SCryptPasswordEncoderTests.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    @Test(expected = IllegalArgumentException.class)
+    public void invalidMemoryCostParameter() {
+        new SCryptPasswordEncoder(UUID.randomUUID().toString(), 2, Integer.MAX_VALUE, 2);     
+    }
+    
+    @Test(expected = IllegalArgumentException.class)
+    public void invalidParallelizationParameter() {
+        new SCryptPasswordEncoder(UUID.randomUUID().toString(), 2, 8, Integer.MAX_VALUE);     
+    }
+
```
Comment (rwinch) [inline crypto/src/test/java/org/springframework/security/crypto/scrypt/SCryptPasswordEncoderTests.java]: Add a test that verifies update of password strength. Pseudocode below

```
SCryptPasswordEncoder oldEncoder = new SCryptPasswordEncoder(... some arguments ...);
SCryptPasswordEncoder newEncoder = new SCryptPasswordEncoder(... all different arguments ...);

String password = "secret";
String oldEncodedPassword = oldEncoder.encode(password);
assertThat(newEncoder.matches(password, oldEncodedPassword)).isTrue();
```

Comment (rwinch) [inline crypto/src/test/java/org/springframework/security/crypto/scrypt/SCryptPasswordEncoderTests.java]: Add a test that encoding the same password twice results in different results (due to a random salt being used).

```java
SCryptPasswordEncoder encoder = new SCryptPasswordEncoder();
String password = "secret";
assertThat(encoder.encode(password)).isNotEqualTo(encoder.encode(secret));
```

**Comment by shazin** [comment]
Pushed the required changes. All Except;

Default parameter values were taken based on the following Scrypt document. 16384, 8, 1, 64

https://www.tarsnap.com/scrypt/scrypt.pdf

As per the guidelines Spaces must be used not tabs. Correct me if I am wrong.

https://github.com/spring-projects/spring-security/blob/master/CONTRIBUTING.md

**Comment by rwinch** [comment]
@shazin Thanks for the response!

> Default parameter values were taken based on the following Scrypt document. 16384, 8, 1, 64

I reviewed the default parameters and they seem to make sense. They use 16mb+ of memory so they should be considered secure.

> As per the guidelines Spaces must be used not tabs. Correct me if I am wrong.

You are right, but the document is wrong! We have switched to tabs :) I updated the document

I went ahead and merged your changes 7d02e259df9e5ac367c237599503e965fd0b4eb6 and applied a bit of polish on it via fc75a679d9e8a8bef4e1d029659637b62b420ce0

Thanks again for your contribution and fast responses!

**Comment by rwinch** [comment]
@shazin Btw...In the release notes I'd like to mention that you added this feature. Do you have a preferred link to use (i.e. link to GitHub profile, Twitter profile, site, etc)

**Comment by shazin** [comment]
Thanks for the final polish and merge. My twitter is https://twitter.com/shazinsadakath

**Pull request closed without merging**